### PR TITLE
[ITG-108]

### PR DESCRIPTION
**Pull request by basiav**
ITG-108 Dodanie do listy zadań wymiennych zadania przechodnio wymienne:
https://kulawska.atlassian.net/browse/ITG-108?atlOrigin=eyJpIjoiMzFkODZkYTE1NTI5NDgzOGJkNzcwODg0OTY1YTQ0NmIiLCJwIjoiaiJ9

How it works - graph and find&union:
1. Select form the relation table all the pairs (task1_id, task2_id) having 'interchangable' relation type.
2.  Create a graph for 'interchangable' relation storage purporses. For each pair of interchangable tasks (task1, task2), create an indirected edge.
3.  Find&union - treat each vertex as an independent set. For each edge, check if its' vertices have a common set representative (parent); if no, then merge the two subsets into one bigger set of tasks that are interchangably related - treat them as interchangably related.
4. Each result set represents tasks that are interchangably related one to every other. As for our result, we pick the one set containing our given task.

### CODE REVIEW

File: bode/bode/utils/graph.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+from collections import defaultdict
+
+
+class FindUnion:
+    POINTER_TO_ONSESELF = -1
```
Comment (john-sonz) [inline bode/bode/utils/graph.py:5]:
```suggestion
    POINTER_TO_ONESELF = -1
```

File: bode/bode/utils/graph.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        find, union = self.find_union.find, self.find_union.union
+
+        for i in self.graph:
+            for j in self.graph[i]:
+                x = find(i)
+                y = find(j)
+                if x != y:
+                    union(x, y)
+
+        return self.find_union.parents
```
Comment (john-sonz) [inline bode/bode/utils/graph.py:53]: `find_union` doesn't need to be attached to `self`
```suggestion
        find_union = FindUnion(uuids)
        find, union = find_union.find, find_union.union

        for i in self.graph:
            for j in self.graph[i]:
                x = find(i)
                y = find(j)
                if x != y:
                    union(x, y)

        return find_union.parents
```

File: bode/bode/utils/graph.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+from collections import defaultdict
+
+
+class FindUnion:
```
Comment (john-sonz) [inline bode/bode/utils/graph.py:4]: What do you think about renaming this to `DisjointSets`? It should read a bit nicer when using it:
```python
disjoint_sets = DisjointSets(x)
disjoint_sets.find(x)
disjoint_sets.union(x, y)
```

File: bode/bode/models/task_relations_actions.py [hunk 4]
```diff
@@ -51,6 +54,54 @@ def get_relation_types(task_id):
...
+        .filter(*filters)
+        .filter(
+            or_(
+                TaskRelation.second_task_id.in_(interchangable_tasks),
+                TaskRelation.first_task_id.in_(interchangable_tasks),
+            )
+        )
+        .join(Task, TaskRelation.second_task_id == Task.id)
+        .filter(Task.id != task_id)
+        .distinct(Task.title)
```
Comment (john-sonz) [inline bode/bode/models/task_relations_actions.py:100]: Doing distinct over `title` might skip some results if there were 2 tasks with the same title in the DB. Any reason why `id` wouldn't work?
```suggestion
        .distinct(Task.id)
```

File: bode/bode/utils/graph.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        elif ranks[x_root] < ranks[y_root]:
+            parents[x_root] = y_root
+        elif x_root != y_root:
+            parents[y_root] = x_root
+            ranks[x_root] += 1
+
+
+class Graph:
+    def __init__(self, V):
+        self.V = V
```
Comment (john-sonz) [inline bode/bode/utils/graph.py:36]: From what I can see the `V` is unused:
```suggestion
    def __init__(self):
```

File: bode/bode/models/task_relations_actions.py [hunk 6]
```diff
@@ -51,6 +54,54 @@ def get_relation_types(task_id):
...
+            uuids.append(str(second_task_id))
+
+        return graph, uuids
+
+    def group_tasks_subsets():
+        task_parent = parents[task_id] if task_id in parents.keys() and parents[task_id] != -1 else task_id
+        tasks_subset = []
+        for key in parents.keys():
+            if task_parent == graph.find_union.find(key):
+                tasks_subset.append(key)
```
Comment (john-sonz) [inline bode/bode/models/task_relations_actions.py:80]:
```suggestion
        tasks_subset = [key for key in parents.keys() if task_parent == graph.find_union.find(key)]
```

**Review by john-sonz — COMMENTED** [review]
💪 POWERFUL 💪 DISJOINT 💪  SETS 💪 WITH 💪 FIND 💪 AND 💪 UNION 💪 FALISZ 💪 WOULD 💪 BE 💪 SO 💪 PROUD

Just some small suggestions, amazing job ❤️

File: bode/bode/models/task_relations_actions.py [hunk 7]
```diff
@@ -51,6 +54,54 @@ def get_relation_types(task_id):
...
 
 
+def get_transitive_interchangable_related_tasks(task_id, filters=list()):
+    from bode.models.task import Task
+
+    def get_distinct_uuids(uuids):
+        return list(set(uuids))
+
+    def get_relation_graph(relations):
+        graph = Graph(len(all_interchangable_tasks))
```
Comment (gregori0o) [inline bode/bode/models/task_relations_actions.py:64]:
```suggestion
        graph = Graph(len(relations))
```

File: bode/bode/utils/graph.py (hunk 2 above)
Comment (gregori0o) [inline bode/bode/utils/graph.py:53]:
```suggestion
    def union_subsets(self, uuids):
        find_union = FindUnion(uuids)
        union = find_union.union

        for i in self.graph:
            for j in self.graph[i]:
                union(i, j)
        return self.find_union.parents
```

You call function find before `union` and in `union`. So those finds are unnecessary.

**Review by gregori0o — COMMENTED** [review]
AMAZING!
Only small suggestion.

I find one problem, but if I.m not mistaken in this sprint is another task, which solve this. Now we can create cycle of interchangeable tasks. We have to block adding relation if two tasks have been already in relation.
Is this task included in task ITG-109? @koszar91

File: bode/bode/models/task_relations_actions.py [hunk 8]
```diff
@@ -51,6 +54,54 @@ def get_relation_types(task_id):
...
+            first_task_id, second_task_id = relation.first_task_id, relation.second_task_id
+            graph.add_edge(str(first_task_id), str(second_task_id))
+            uuids.append(str(first_task_id))
+            uuids.append(str(second_task_id))
+
+        return graph, uuids
+
+    def group_tasks_subsets():
+        task_parent = parents[task_id] if task_id in parents.keys() and parents[task_id] != -1 else task_id
+        tasks_subset = []
```
Comment (gregori0o) [inline bode/bode/models/task_relations_actions.py:77]:
```suggestion
    def group_tasks_subsets():
        task_parent = graph.find_union.find(task_id) if task_id in parents.keys() else task_id
        tasks_subset = []
```

We can't be sure that `parents[task_id]` is root of this task.

**Comment by koszar91** [comment]
> I didn't think about it, but actually it can be done in ITG-109, I'll consider it, thanks!



> AMAZING! Only small suggestion.
> 
> I find one problem, but if I.m not mistaken in this sprint is another task, which solve this. Now we can create cycle of interchangeable tasks. We have to block adding relation if two tasks have been already in relation. Is this task included in task ITG-109? @koszar91

File: bode/bode/utils/graph.py (hunk 2 above)
Comment (basiav) [reply bode/bode/utils/graph.py:53]: it seems to need to be attached, graph's property `find_union` (name updated to `disjoint_sets`) is called in a few lines in tasks_relations_actions.py

**Review by john-sonz — APPROVED** [review]
💜  ❤️ 💚  💛  💙

**Pull request merged**